AX: When running tests, force default accessibility settings to avoid failures due to the machine's local settings

twilco opened the pull request:
#### 71884ba628172c0ad09a16ee45cd22865a749a90
<pre>
AX: When running tests, force default accessibility settings to avoid failures due to the machine&apos;s local settings
<a href="https://bugs.webkit.org/show_bug.cgi?id=298897">https://bugs.webkit.org/show_bug.cgi?id=298897</a>
<a href="https://rdar.apple.com/148661873">rdar://148661873</a>

Reviewed by Joshua Hoffman.

Add a new ForceDefaultAccessibilitySettings defaults value and read it in new AXPreferenceHelpers methods to avoid
calling into accessibility frameworks for the system-local setting value.

* Source/WebKit/Shared/AccessibilityPreferences.h:
* Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm: Added.
(AXPreferenceHelpers::shouldUseDefault):
(AXPreferenceHelpers::reduceMotionEnabled):
(AXPreferenceHelpers::increaseButtonLegibility):
(AXPreferenceHelpers::enhanceTextLegibility):
(AXPreferenceHelpers::darkenSystemColors):
(AXPreferenceHelpers::invertColorsEnabled):
(AXPreferenceHelpers::imageAnimationEnabled):
(AXPreferenceHelpers::enhanceTextLegibilityOverall):
(AXPreferenceHelpers::prefersNonBlinkingCursor):
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _allowAnimationControls]):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::accessibilityPreferences):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/mac/main.mm:
(setDefaultsToConsistentValuesForTesting):

Canonical link: <a href="https://commits.webkit.org/300007@main">https://commits.webkit.org/300007@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/560457fa62811cd92c1c19793ba4b561d3528915

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121041 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40737 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127459 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73121 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b5bd3177-abff-4aa7-9430-f4d5f6402f79) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122917 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41439 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49316 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91932 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/aefe42e9-0ec4-4215-a975-27f3ffc7fb82) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123993 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33082 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108500 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72620 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a7ee1a04-1a4a-443f-ab79-874b26445e6d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32105 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26603 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71049 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102587 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26782 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130313 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47968 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36448 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100541 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48336 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104667 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100443 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25458 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45848 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23900 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44656 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47826 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53539 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47297 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50644 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48981 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm:
```diff
@@ -0,0 +1,116 @@
+#if ENABLE(ACCESSIBILITY_ANIMATION_CONTROL)
+    if (shouldUseDefault()) [[unlikely]]
+        return WebKit::defaultImageAnimationEnabled;
+    if (auto* functionPointer = _AXSReduceMotionAutoplayAnimatedImagesEnabledPtr())
```
do we still need to use this Ptr method?

twilco replied:
Maybe not, but not 100% sure considering WebKit is backported some releases. Let's change this in a later PR since I just copy-pasted this from its old location.

fleizach commented on Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm:
```diff
@@ -0,0 +1,116 @@
+{
+    if (shouldUseDefault()) [[unlikely]]
+        return WebKit::defaultPerAppSettingsState;
+    auto appId = applicationBundleIdentifier().createCFString();
```
I'm not sure what auto creates here but it seems like this createCFString() will leak

lilyspiniolas replied:
It creates a `RetainPtr<CFStringRef>` (the style guide states that `auto` should not be used for smart pointers, though)

twilco replied:
I'll change it to be `RetainPtr` 👍

lilyspiniolas commented on Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm:
```diff
@@ -0,0 +1,116 @@
+
+static bool shouldUseDefault()
+{
+    NSNumber *forceDefault = [[NSUserDefaults standardUserDefaults] objectForKey:@"ForceDefaultAccessibilitySettings"];
```
```suggestion
    RetainPtr forceDefault = [[NSUserDefaults standardUserDefaults] objectForKey:@"ForceDefaultAccessibilitySettings"];
```

twilco replied:
Addressed all of these in latest commit!

lilyspiniolas commented on Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm:
```diff
@@ -0,0 +1,116 @@
+{
+    if (shouldUseDefault()) [[unlikely]]
+        return WebKit::defaultPerAppSettingsState;
+    auto appId = applicationBundleIdentifier().createCFString();
```
```suggestion
    RetainPtr appId = applicationBundleIdentifier().createCFString();
```

lilyspiniolas commented on Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm:
```diff
@@ -0,0 +1,116 @@
+{
+    if (shouldUseDefault()) [[unlikely]]
+        return WebKit::defaultPerAppSettingsState;
+    auto appId = applicationBundleIdentifier().createCFString();
```
```suggestion
    RetainPtr appId = applicationBundleIdentifier().createCFString();
```

lilyspiniolas commented on Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm:
```diff
@@ -0,0 +1,116 @@
+{
+    if (shouldUseDefault()) [[unlikely]]
+        return WebKit::defaultPerAppSettingsState;
+    auto appId = applicationBundleIdentifier().createCFString();
```
```suggestion
    RetainPtr appId = applicationBundleIdentifier().createCFString();
```

lilyspiniolas commented on Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm:
```diff
@@ -0,0 +1,116 @@
+{
+    if (shouldUseDefault()) [[unlikely]]
+        return WebKit::defaultPerAppSettingsState;
+    auto appId = applicationBundleIdentifier().createCFString();
```
```suggestion
    RetainPtr appId = applicationBundleIdentifier().createCFString();
```

lilyspiniolas commented on Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm:
```diff
@@ -0,0 +1,116 @@
+#if ENABLE(ACCESSIBILITY_ANIMATION_CONTROL)
+    if (shouldUseDefault()) [[unlikely]]
+        return WebKit::defaultImageAnimationEnabled;
+    if (auto* functionPointer = _AXSReduceMotionAutoplayAnimatedImagesEnabledPtr())
```
```suggestion
    if (RetainPtr functionPointer = _AXSReduceMotionAutoplayAnimatedImagesEnabledPtr())
```

minorninth commented on Source/WebKit/Shared/AccessibilityPreferences.h:
```diff
@@ -76,20 +76,44 @@ inline AXValueState fromWebKitAXValueState(WebKitAXValueState value)
 }
 #endif
 
+inline constexpr bool defaultImageAnimationEnabled = true;
```
The word "default" is so overloaded I wonder if we should pick a different word for this, like factory, initial, baseline, out of box.

Or we could lean into the testing use case and call this initialImageAnimationEnabledForTests

twilco replied:
Fixed in latest commit!

twilco committed:
AX: When running tests, force default accessibility settings to avoid failures due to the machine's local settings
https://bugs.webkit.org/show_bug.cgi?id=298897
rdar://148661873

Reviewed by Joshua Hoffman.

Add a new ForceDefaultAccessibilitySettings defaults value and read it in new AXPreferenceHelpers methods to avoid
calling into accessibility frameworks for the system-local setting value.

* Source/WebKit/Shared/AccessibilityPreferences.h:
* Source/WebKit/Shared/Cocoa/AccessibilityPreferences.mm: Added.
(AXPreferenceHelpers::shouldUseDefault):
(AXPreferenceHelpers::reduceMotionEnabled):
(AXPreferenceHelpers::increaseButtonLegibility):
(AXPreferenceHelpers::enhanceTextLegibility):
(AXPreferenceHelpers::darkenSystemColors):
(AXPreferenceHelpers::invertColorsEnabled):
(AXPreferenceHelpers::imageAnimationEnabled):
(AXPreferenceHelpers::enhanceTextLegibilityOverall):
(AXPreferenceHelpers::prefersNonBlinkingCursor):
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _allowAnimationControls]):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::accessibilityPreferences):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/mac/main.mm:
(setDefaultsToConsistentValuesForTesting):

Canonical link: https://commits.webkit.org/300007@main
